File: ureka-framework-main/tests/conftest.py
import inspect
from ureka_framework.resource.logger.simple_logger import simple_log

from typing import Tuple
from ureka_framework.logic.device_controller import DeviceController
from ureka_framework.model.message_model import u_ticket
import ureka_framework.model.data_model.this_device as this_device
from ureka_framework.resource.crypto.serialization_util import dict_to_jsonstr


######################################################
# Helper Functions
######################################################
def get_current_class_name() -> str:
    # Get Current Class Name
    for frame_info in inspect.stack():
        frame_locals = frame_info.frame.f_locals
        if "self" in frame_locals:
            return frame_locals["self"].__class__.__name__


def get_current_function_name() -> str:
    # Get Current Function Name called by the test
    for frame_info in inspect.stack():
        frame_locals = frame_info.frame.f_locals
        if "self" in frame_locals:
            return frame_info.function


def current_setup_log() -> None:
    # Log
    simple_log("info", "")
    if get_current_class_name() != None:
        simple_log("info", "*" * 100)
        simple_log("info", f"Setup: {get_current_class_name()}")
        simple_log("info", "*" * 100)


def current_test_given_log() -> None:
    # Log
    if (
        get_current_function_name() != "_hookexec"
        and get_current_function_name() != None
    ):
        simple_log("info", "*" * 50)
        simple_log("info", f"Given: {get_current_function_name()}")
        simple_log("info", "*" * 50)


def current_test_when_and_then_log() -> None:
    # Log
    if (
        get_current_function_name() != "_hookexec"
        and get_current_function_name() != None
    ):
        simple_log("info", "*" * 50)
        simple_log("info", f"When & Then: {get_current_function_name()}")
        simple_log("info", "*" * 50)


def current_teardown_log() -> None:
    # Log
    if get_current_class_name() != None:
        simple_log("info", "*" * 100)
        simple_log("info", f"Teardown: {get_current_class_name()}")
        simple_log("info", "*" * 100)


######################################################
# Helper Functions (Simulated Comm)
######################################################
def create_simulated_comm_connection(end1: DeviceController, end2: DeviceController):
    # # Re-open the sender/receiver
    # end1.msg_sender.start_simulated_comm()
    # end2.msg_sender.start_simulated_comm()

    # Set Sender (on Main Thread) & Start Reciever Thread
    end1.msg_receiver.create_simulated_comm_connection(end2)
    end2.msg_receiver.create_simulated_comm_connection(end1)

    simple_log(
        "info",
        f"+ Connection between {end1.shared_data.this_device.device_name} and {end2.shared_data.this_device.device_name} is started...",
    )


def wait_simulated_comm_completed(end1: DeviceController, end2: DeviceController):
    # Wait for all sender/receiver to finish their works (block last 1st make log beautiful)
    end1.msg_sender.wait_simulated_comm_completed()
    end2.msg_sender.wait_simulated_comm_completed()

    simple_log(
        "info",
        f"+ Connection between {end1.shared_data.this_device.device_name} and {end2.shared_data.this_device.device_name} is completed...",
    )
    simple_log("info", "")
    simple_log("info", "")
    simple_log("info", "")


######################################################
# Helper Functions (Reusable Test Data)
######################################################
def device_manufacturer_server() -> DeviceController:
    # GIVEN: Initialized DM's CS
    cloud_server_dm = DeviceController(
        device_type=this_device.USER_AGENT_OR_CLOUD_SERVER,
        device_name="cloud_server_dm",
    )
    cloud_server_dm.executor._execute_one_time_intialize_agent_or_server()

    return cloud_server_dm


def device_manufacturer_server_and_her_device() -> (
    Tuple[DeviceController, DeviceController]
):
    # GIVEN: Initialized DM's CS
    cloud_server_dm = device_manufacturer_server()

    # GIVEN: Uninitialized IoTD
    iot_device = DeviceController(
        device_type=this_device.IOT_DEVICE,
        device_name="iot_device",
    )

    # WHEN: Issuer: DM's CS generate & send the intialization_u_ticket to Uninitialized IoTD
    create_simulated_comm_connection(cloud_server_dm, iot_device)
    id_for_initialization_u_ticket = "no_id"
    generated_request: dict = {
        "device_id": f"{id_for_initialization_u_ticket}",
        "holder_id": f"{cloud_server_dm.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_INITIALIZATION_UTICKET}",
    }
    cloud_server_dm.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_herself(
        device_id=id_for_initialization_u_ticket, arbitrary_dict=generated_request
    )
    cloud_server_dm.flow_apply_u_ticket.holder_apply_u_ticket(
        id_for_initialization_u_ticket
    )
    wait_simulated_comm_completed(cloud_server_dm, iot_device)

    return (cloud_server_dm, iot_device)


def device_owner_agent() -> DeviceController:
    # GIVEN: Initialized DM's CS
    user_agent_do = DeviceController(
        device_type=this_device.USER_AGENT_OR_CLOUD_SERVER,
        device_name="user_agent_do",
    )
    user_agent_do.executor._execute_one_time_intialize_agent_or_server()

    return user_agent_do


def device_owner_agent_and_her_device() -> Tuple[DeviceController, DeviceController]:
    # GIVEN: Initialized DM's CS and DM's IoTD
    (
        cloud_server_dm,
        iot_device,
    ) = device_manufacturer_server_and_her_device()

    # GIVEN: Initialized DO's UA
    user_agent_do = device_owner_agent()

    # WHEN: Issuer: DM's CS generate & send the ownership_u_ticket to DO's UA
    create_simulated_comm_connection(cloud_server_dm, user_agent_do)
    target_device_id = iot_device.shared_data.this_device.device_pub_key_str
    generated_request: dict = {
        "device_id": f"{target_device_id}",
        "holder_id": f"{user_agent_do.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_OWNERSHIP_UTICKET}",
    }
    cloud_server_dm.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_holder(
        device_id=target_device_id, arbitrary_dict=generated_request
    )
    wait_simulated_comm_completed(user_agent_do, cloud_server_dm)

    # WHEN: Holder: DO's UA forward the ownership_u_ticket
    create_simulated_comm_connection(user_agent_do, iot_device)
    user_agent_do.flow_apply_u_ticket.holder_apply_u_ticket(target_device_id)
    wait_simulated_comm_completed(user_agent_do, iot_device)

    return (user_agent_do, iot_device)


def device_owner_agent_and_her_session() -> Tuple[DeviceController, DeviceController]:
    # GIVEN: Initialized DO's UA and DO's IoTD
    (
        user_agent_do,
        iot_device,
    ) = device_owner_agent_and_her_device()

    # WHEN:
    current_test_when_and_then_log()

    # WHEN: Issuer: DO's UA generate the self_access_u_ticket to herself
    target_device_id = iot_device.shared_data.this_device.device_pub_key_str
    generated_task_scope = dict_to_jsonstr({"ALL": "allow"})
    generated_request: dict = {
        "device_id": f"{target_device_id}",
        "holder_id": f"{user_agent_do.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_SELFACCESS_UTICKET}",
        "task_scope": f"{generated_task_scope}",
    }
    user_agent_do.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_herself(
        device_id=target_device_id, arbitrary_dict=generated_request
    )

    # WHEN: Holder: DO's UA forward the self_access_u_ticket
    create_simulated_comm_connection(user_agent_do, iot_device)
    generated_command = "HELLO-1"
    user_agent_do.flow_apply_u_ticket.holder_apply_u_ticket(
        target_device_id, generated_command
    )
    wait_simulated_comm_completed(user_agent_do, iot_device)

    return (user_agent_do, iot_device)


def enterprise_provider_server() -> DeviceController:
    # GIVEN: Initialized EP's CS
    cloud_server_ep = DeviceController(
        device_type=this_device.USER_AGENT_OR_CLOUD_SERVER,
        device_name="cloud_server_ep",
    )
    cloud_server_ep.executor._execute_one_time_intialize_agent_or_server()

    return cloud_server_ep


def enterprise_provider_server_and_her_session() -> (
    Tuple[DeviceController, DeviceController, DeviceController]
):
    # GIVEN: Initialized DO's UA and DO's IoTD
    (
        user_agent_do,
        iot_device,
    ) = device_owner_agent_and_her_device()

    # GIVEN: Initialized EP's CS
    cloud_server_ep = enterprise_provider_server()

    # WHEN: Issuer: DO's UA generate & send the access_u_ticket to EP's CS
    create_simulated_comm_connection(user_agent_do, cloud_server_ep)
    target_device_id = iot_device.shared_data.this_device.device_pub_key_str
    generated_task_scope = dict_to_jsonstr({"ALL": "allow"})
    generated_request: dict = {
        "device_id": f"{target_device_id}",
        "holder_id": f"{cloud_server_ep.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_ACCESS_UTICKET}",
        "task_scope": f"{generated_task_scope}",
    }
    generated_request: dict = {
        "device_id": f"{target_device_id}",
        "holder_id": f"{cloud_server_ep.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_ACCESS_UTICKET}",
        "task_scope": f"{generated_task_scope}",
    }
    user_agent_do.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_holder(
        device_id=target_device_id, arbitrary_dict=generated_request
    )
    wait_simulated_comm_completed(cloud_server_ep, user_agent_do)

    # WHEN: Holder: EP's CS forward the access_u_ticket
    create_simulated_comm_connection(cloud_server_ep, iot_device)
    generated_command = "HELLO-1"
    cloud_server_ep.flow_apply_u_ticket.holder_apply_u_ticket(
        target_device_id, generated_command
    )
    wait_simulated_comm_completed(cloud_server_ep, iot_device)

    return (user_agent_do, cloud_server_ep, iot_device)


def enterprise_provider_server_and_her_limited_session() -> (
    Tuple[DeviceController, DeviceController, DeviceController]
):
    # GIVEN: Initialized DO's UA and DO's IoTD
    (
        user_agent_do,
        iot_device,
    ) = device_owner_agent_and_her_device()

    # GIVEN: Initialized EP's CS
    cloud_server_ep = enterprise_provider_server()

    # WHEN: Issuer: DO's UA generate & send the access_u_ticket to EP's CS
    create_simulated_comm_connection(user_agent_do, cloud_server_ep)
    target_device_id = iot_device.shared_data.this_device.device_pub_key_str
    generated_task_scope = dict_to_jsonstr(
        {
            "SAY-HELLO-1": "allow",
            "SAY-HELLO-2": "allow",
            "SAY-HELLO-3": "forbid",
        }
    )
    generated_request: dict = {
        "device_id": f"{target_device_id}",
        "holder_id": f"{cloud_server_ep.shared_data.this_person.person_pub_key_str}",
        "u_ticket_type": f"{u_ticket.TYPE_ACCESS_UTICKET}",
        "task_scope": f"{generated_task_scope}",
    }
    user_agent_do.flow_issuer_issue_u_ticket.issuer_issue_u_ticket_to_holder(
        device_id=target_device_id, arbitrary_dict=generated_request
    )
    wait_simulated_comm_completed(cloud_server_ep, user_agent_do)

    # WHEN: Holder: EP's CS forward the access_u_ticket
    create_simulated_comm_connection(cloud_server_ep, iot_device)
    generated_command = "HELLO-1"
    cloud_server_ep.flow_apply_u_ticket.holder_apply_u_ticket(
        target_device_id, generated_command
    )
    wait_simulated_comm_completed(cloud_server_ep, iot_device)

    return (user_agent_do, cloud_server_ep, iot_device)


def attacker_server() -> DeviceController:
    # GIVEN: Initialized ATK's CS
    cloud_server_atk = DeviceController(
        device_type=this_device.USER_AGENT_OR_CLOUD_SERVER,
        device_name="cloud_server_atk",
    )
    cloud_server_atk.executor._execute_one_time_intialize_agent_or_server()

    return cloud_server_atk


######################################################
# Fixtures (Reusable Test Data without Logging)
######################################################

File: UrekaApp/app/src/androidTest/java/Conftest.java
import java.util.HashMap;
import java.util.Map;

import ureka.framework.logic.DeviceController;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.Pair;
import ureka.framework.resource.Triple;
import ureka.framework.resource.crypto.SerializationUtil;
import ureka.framework.resource.logger.SimpleLogger;

public class Conftest {
    // Helper Functions

    // Get Current Class Name
    public static String getCurrentClassName() {
        try {
            return new Throwable().getStackTrace()[1].getClassName();
        } catch (Exception e) {
            SimpleLogger.simpleLog("warning","Failed to get class name.");
            return null;
        }
    }

    // Get Current Function Name called by the test
    public static String getCurrentFunctionName() {
        try {
            return new Throwable().getStackTrace()[1].getMethodName();
        } catch (Exception e) {
            SimpleLogger.simpleLog("warning","Failed to get function name.");
            return null;
        }
    }

    // Log
    public static void currentSetupLog() {
        SimpleLogger.simpleLog("info","");
        String className = getCurrentClassName();
        if (className != null) {
            SimpleLogger.simpleLog("info","*".repeat(100));
            SimpleLogger.simpleLog("info","Setup: " + className);
            SimpleLogger.simpleLog("info","*".repeat(100));
        }
    }

    // Log
    public static void currentTestGivenLog() {
        String functionName = getCurrentFunctionName();
        if (!"_hookexec".equals(functionName) && functionName != null) {
            SimpleLogger.simpleLog("info","*".repeat(50));
            SimpleLogger.simpleLog("info","Given: " + functionName);
            SimpleLogger.simpleLog("info","*".repeat(50));
        }
    }

    // Log
    public static void currentTestWhenAndThenLog() {
        String functionName = getCurrentFunctionName();
        if (!"_hookexec".equals(functionName) && functionName != null) {
            SimpleLogger.simpleLog("info","*".repeat(50));
            SimpleLogger.simpleLog("info","When & Then: " + functionName);
            SimpleLogger.simpleLog("info","*".repeat(50));
        }
    }

    // Log
    public static void currentTeardownLog() {
        String className = getCurrentClassName();
        if (className != null) {
            SimpleLogger.simpleLog("info","*".repeat(100));
            SimpleLogger.simpleLog("info","Teardown: " + className);
            SimpleLogger.simpleLog("info","*".repeat(100));
        }
    }

    // Helper Functions (Simulated Comm)
//    public static void createSimulatedCommConnection(DeviceController end1, DeviceController end2) {
//        // Set Sender (on Main Thread) & Start Receiver Thread
//        end1.getMsgReceiver().createSimulatedCommConnection(end2);
//        end2.getMsgReceiver().createSimulatedCommConnection(end1);
//
//        SimpleLogger.simpleLog("info", String.format("Connection between %s and %s is started...",
//                end1.getSharedData().getThisDevice().getDeviceName(),
//                end2.getSharedData().getThisDevice().getDeviceName()));
//    }
//
//    // Wait for all sender/receiver to finish their works (block last 1st make log beautiful)
//    public static void waitSimulatedCommCompleted(DeviceController end1, DeviceController end2) {
//        try {
//            end1.getMsgSender().waitSimulatedCommCompleted();
//            end2.getMsgSender().waitSimulatedCommCompleted();
//        } catch (InterruptedException e) {
//            throw new RuntimeException(e);
//        }
//
//        SimpleLogger.simpleLog("info", String.format("Connection between %s and %s is completed...",
//                end1.getSharedData().getThisDevice().getDeviceName(),
//                end2.getSharedData().getThisDevice().getDeviceName()));
//        SimpleLogger.simpleLog("info","");
//        SimpleLogger.simpleLog("info","");
//        SimpleLogger.simpleLog("info","");
//    }

    // Helper Functions (Reusable Test Data)
    public static DeviceController deviceManufacturerServer() {
        DeviceController cloudServerDM = new DeviceController(ThisDevice.USER_AGENT_OR_CLOUD_SERVER, "cloudServerDM");
        cloudServerDM.getExecutor()._executeOneTimeInitializeAgentOrServer();

        return cloudServerDM;
    }

    public static Pair deviceManufacturerServerAndHerDevice() {
        // GIVEN: Initialized DM's CS
        DeviceController cloudServerDM = deviceManufacturerServer();

        // GIVEN: Uninitialized IoTD
        DeviceController iotDevice = new DeviceController(ThisDevice.IOT_DEVICE, "iotDevice");

        // WHEN: Issuer: DM's CS generate & send the initialization_u_ticket to Uninitialized IoTD
        // createSimulatedCommConnection(cloudServerDM, iotDevice);
        String idForInitializationUTicket = "noId";
        Map<String, String> generatedRequest = new HashMap<>();
        generatedRequest.put("device_id", idForInitializationUTicket);
        generatedRequest.put("holder_id", cloudServerDM.getSharedData().getThisPerson().getPersonPubKeyStr());
        generatedRequest.put("u_ticket_type", UTicket.TYPE_INITIALIZATION_UTICKET);

        cloudServerDM.getFlowIssuerIssueUTicket().issuerIssueUTicketToHerself(idForInitializationUTicket, generatedRequest);

        cloudServerDM.getFlowApplyUTicket().holderApplyUTicket(idForInitializationUTicket);
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        cloudServerDM.getMsgReceiver()._recvXxxMessage();

        // waitSimulatedCommCompleted(cloudServerDM, iotDevice);

        return new Pair(cloudServerDM, iotDevice);
    }

    public static DeviceController deviceOwnerAgent() {
        // GIVEN: Initialized DM's CS
        DeviceController userAgentDO = new DeviceController(ThisDevice.USER_AGENT_OR_CLOUD_SERVER, "userAgentDO");
        userAgentDO.getExecutor()._executeOneTimeInitializeAgentOrServer();

        return userAgentDO;
    }

    public static Pair deviceOwnerAgentAndHerDevice() {
        // GIVEN: Initialized DM's CS and DM's IoTD
        Pair cloudServerDMAndIoT = deviceManufacturerServerAndHerDevice();
        DeviceController cloudServerDM = (DeviceController) cloudServerDMAndIoT.getPairFirst();
        DeviceController iotDevice = (DeviceController) cloudServerDMAndIoT.getPairSecond();

        // GIVEN: Initialized DO's UA
        DeviceController userAgentDO = deviceOwnerAgent();

        // WHEN: Issuer: DM's CS generate & send the ownership_u_ticket to DO's UA
        // createSimulatedCommConnection(cloudServerDM, userAgentDO);
        String targetDeviceId = iotDevice.getSharedData().getThisDevice().getDevicePubKeyStr();
        Map<String, String> generatedRequest = new HashMap<>();
        generatedRequest.put("device_id", targetDeviceId);
        generatedRequest.put("holder_id", userAgentDO.getSharedData().getThisPerson().getPersonPubKeyStr());
        generatedRequest.put("u_ticket_type", UTicket.TYPE_OWNERSHIP_UTICKET);

        // cloudServerDM.getFlowIssuerIssueUTicket().issuerIssueUTicketToHolder(targetDeviceId, generatedRequest);
//        userAgentDO.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(userAgentDO, cloudServerDM);

        // WHEN: Holder: DO's UA forward the ownershipUTicket
        // createSimulatedCommConnection(userAgentDO, iotDevice);
        userAgentDO.getFlowApplyUTicket().holderApplyUTicket(targetDeviceId);
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        userAgentDO.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(userAgentDO, iotDevice);

        return new Pair(userAgentDO, iotDevice);
    }

    public static Pair deviceOwnerAgentAndHerSession() {
        // GIVEN: Initialized DO's UA and DO's IoTD
        Pair deviceTuple = deviceOwnerAgentAndHerDevice();
        DeviceController userAgentDO = (DeviceController) deviceTuple.getPairFirst();
        DeviceController iotDevice = (DeviceController) deviceTuple.getPairSecond();

        // WHEN:
        currentTestWhenAndThenLog();

        // WHEN: Issuer: DO's UA generate the self_access_u_ticket to herself
        String targetDeviceId = iotDevice.getSharedData().getThisDevice().getDevicePubKeyStr();
        String generatedTaskScope = SerializationUtil.dictToJsonStr(Map.of("ALL", "allow"));
        Map<String, String> generatedRequest = new HashMap<>();
        generatedRequest.put("device_id", targetDeviceId);
        generatedRequest.put("holder_id", userAgentDO.getSharedData().getThisPerson().getPersonPubKeyStr());
        generatedRequest.put("u_ticket_type", UTicket.TYPE_SELFACCESS_UTICKET);
        generatedRequest.put("task_scope", generatedTaskScope);

        userAgentDO.getFlowIssuerIssueUTicket().issuerIssueUTicketToHerself(targetDeviceId, generatedRequest);

        // WHEN: Holder: DO's UA forward the self_access_u_ticket
        // createSimulatedCommConnection(userAgentDO, iotDevice);
        String generatedCommand = "HELLO-1";
        userAgentDO.getFlowApplyUTicket().holderApplyUTicket(targetDeviceId, generatedCommand);
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        userAgentDO.getMsgReceiver()._recvXxxMessage();
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        userAgentDO.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(userAgentDO, iotDevice);

        return new Pair(userAgentDO, iotDevice);
    }

    // GIVEN: Initialized EP's CS
    public static DeviceController enterpriseProviderServer() {
        DeviceController cloudServerEP = new DeviceController(ThisDevice.USER_AGENT_OR_CLOUD_SERVER, "cloudServerEp");
        cloudServerEP.getExecutor()._executeOneTimeInitializeAgentOrServer();

        return cloudServerEP;
    }

    public static Triple enterpriseProviderServerAndHerSession() {
        // GIVEN: Initialized DO's UA and DO's IoTD
        Pair deviceTuple = deviceOwnerAgentAndHerDevice();
        DeviceController userAgentDO = (DeviceController) deviceTuple.getPairFirst();
        DeviceController iotDevice = (DeviceController) deviceTuple.getPairSecond();

        // GIVEN: Initialized EP's CS
        DeviceController cloudServerEP = enterpriseProviderServer();

        // WHEN: Issuer: DO's UA generate & send the access_u_ticket to EP's CS
        // createSimulatedCommConnection(userAgentDO, cloudServerEP);
        String targetDeviceId = iotDevice.getSharedData().getThisDevice().getDevicePubKeyStr();
        String generatedTaskScope = SerializationUtil.dictToJsonStr(Map.of("ALL", "allow"));
        Map<String, String> generatedRequest = new HashMap<>();
        generatedRequest.put("device_id", targetDeviceId);
        generatedRequest.put("holder_id", cloudServerEP.getSharedData().getThisPerson().getPersonPubKeyStr());
        generatedRequest.put("u_ticket_type", UTicket.TYPE_ACCESS_UTICKET);
        generatedRequest.put("task_scope", generatedTaskScope);

        // userAgentDO.getFlowIssuerIssueUTicket().issuerIssueUTicketToHolder(targetDeviceId, generatedRequest);
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(cloudServerEP, userAgentDO);

        // WHEN: Holder: EP's CS forward the access_u_ticket
        // createSimulatedCommConnection(cloudServerEP, iotDevice);
        String generatedCommand = "HELLO-1";
        cloudServerEP.getFlowApplyUTicket().holderApplyUTicket(targetDeviceId, generatedCommand);
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(cloudServerEP, iotDevice);

        return new Triple(userAgentDO, cloudServerEP, iotDevice);
    }

    public static Triple enterpriseProviderServerAndHerLimitedSession() {
        // GIVEN: Initialized DO's UA and DO's IoTD
        Pair deviceTuple = deviceOwnerAgentAndHerDevice();
        DeviceController userAgentDO = (DeviceController) deviceTuple.getPairFirst();
        DeviceController iotDevice = (DeviceController) deviceTuple.getPairSecond();

        // GIVEN: Initialized EP's CS
        DeviceController cloudServerEP = enterpriseProviderServer();

        // WHEN: Issuer: DO's UA generate & send the access_u_ticket to EP's CS
        // createSimulatedCommConnection(userAgentDO, cloudServerEP);
        String targetDeviceId = iotDevice.getSharedData().getThisDevice().getDevicePubKeyStr();
        String generatedTaskScope = SerializationUtil.dictToJsonStr(Map.of(
                "SAY-HELLO-1", "allow",
                "SAY-HELLO-2", "allow",
                "SAY-HELLO-3", "forbid"
        ));
        Map<String, String> generatedRequest = new HashMap<>();
        generatedRequest.put("device_id", targetDeviceId);
        generatedRequest.put("holder_id", cloudServerEP.getSharedData().getThisPerson().getPersonPubKeyStr());
        generatedRequest.put("u_ticket_type", UTicket.TYPE_ACCESS_UTICKET);
        generatedRequest.put("task_scope", generatedTaskScope);

        // userAgentDO.getFlowIssuerIssueUTicket().issuerIssueUTicketToHolder(targetDeviceId, generatedRequest);
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(cloudServerEP, userAgentDO);

        // WHEN: Holder: EP's CS forward the access_u_ticket
        // createSimulatedCommConnection(cloudServerEP, iotDevice);
        String generatedCommand = "HELLO-1";
        cloudServerEP.getFlowApplyUTicket().holderApplyUTicket(targetDeviceId, generatedCommand);
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
//        iotDevice.getMsgReceiver()._recvXxxMessage();
//        cloudServerEP.getMsgReceiver()._recvXxxMessage();
        // waitSimulatedCommCompleted(cloudServerEP, iotDevice);

        return new Triple(userAgentDO, cloudServerEP, iotDevice);
    }

    public static DeviceController attackerServer() {
        // GIVEN: Initialized ATK's CS
        DeviceController cloudServerATK = new DeviceController(ThisDevice.USER_AGENT_OR_CLOUD_SERVER, "cloudServerATK");
        cloudServerATK.getExecutor()._executeOneTimeInitializeAgentOrServer();

        return cloudServerATK;
    }

    // Fixture
}
